src/lib/logger/logger.ml

```diff
@@ -81,11 +81,160 @@ module Message = struct
     List.for_all refs ~f:(Metadata.mem t.metadata)
 end
 
-type t = {null: bool; metadata: Metadata.t}
+module Processor = struct
+  module type S = sig
+    type t
+
+    val process : t -> Message.t -> string option
+  end
+
+  type t = T : (module S with type t = 't) * 't -> t
+
+  module Raw = struct
+    type t = unit
+
+    let create () = ()
+
+    let process () msg = Some (Yojson.Safe.to_string (Message.to_yojson msg))
+  end
+
+  module Pretty = struct
+    type t = {log_level: Level.t; config: Logproc_lib.Interpolator.config}
+
+    let create ~log_level ~config = {log_level; config}
+
+    let process {log_level; config} msg =
+      let open Message in
+      if msg.level < log_level then None
+      else
+        match
+          Logproc_lib.Interpolator.interpolate config msg.message msg.metadata
+        with
+        | Error err ->
+            Core.printf "logproc interpolation error: %s\n" err ;
+            None
+        | Ok (str, _) ->
+            Some (Level.show msg.level ^ ": " ^ str)
+  end
+
+  let raw () = T ((module Raw), Raw.create ())
+
+  let pretty ~log_level ~config =
+    T ((module Pretty), Pretty.create ~log_level ~config)
+end
+
+module Transport = struct
+  module type S = sig
+    type t
+
+    val transport : t -> string -> unit
+  end
+
+  type t = T : (module S with type t = 't) * 't -> t
+
+  module Stdout = struct
+    type t = unit
+
+    let create () = ()
+
+    let transport () = Core.print_endline
+  end
+
+  let stdout () = T ((module Stdout), Stdout.create ())
+
+  module File_system = struct
+    module Dumb_logrotate = struct
+      open Core.Unix
+
+      (* 512MB *)
+      let max_size = 1024 * 1024 * 512
+
+      let log_perm = 0o644
+
+      let primary_log_name = "coda.log.0"
+
+      let secondary_log_name = "coda.log.1"
+
+      type t =
+        { directory: string
+        ; mutable primary_log: File_descr.t
+        ; mutable primary_log_size: int }
+
+      let create ~directory =
+        let primary_log_loc = Filename.concat directory primary_log_name in
+        let primary_log_size =
+          if Result.is_ok (access primary_log_loc [`Read; `Write]) then
+            let log_stats = stat primary_log_loc in
+            Int64.to_int_exn log_stats.st_size
+          else 0
+        in
+        let primary_log =
+          openfile ~perm:log_perm ~mode:[O_RDWR; O_APPEND] primary_log_loc
```

Comment:
If this fails, I think the UX right now is quite bad. I believe it'll throw an exception? Also, is this intentionally avoiding `Asynx.Unix` in favor of `Core.Unix`? I think it's also noteworthy that this will append to an existing log file. Not sure how common that is in our workflows - but I think that during a "dirty shutdown" (incl most kinds of exceptions?) the log file could possibly be truncated without a complete JSON line being written. If the daemon gets autorestarted (which we wanted to start doing at some point) it'll keep writing to the corrupted log file. Not sure how big of a problem this is - logproc is smart enough to skip over a corrupt line but other tools might not be.

Reply:
(Async vs Core is explained in another comment)

Right now, if this fails, it will throw an exception. However, it shouldn't fail as long as the config directory exists and we have permissions to it. If either of those two aren't true, the code would already crash from an exception before it reaches this code, so it's not a regression at least.

As for what to do with existing logs: I went with the easiest solution for now which is to just continue on top of the logs that are already there. I didn't want to delete the older log if we don't have to for now. Since these logs are primarily meant for testers to send us when they have crashes, I'm not too concerned about the possiblity of malformed lines since it's easy to identify and remove those lines (and as you mentioned, logproc already handles this). I will make an issue for revisiting this UX though.